exporter/awsemfexporter/internal/appsignals/useragent.go

```diff
@@ -0,0 +1,116 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package appsignals // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter/internal/appsignals"
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/jellydator/ttlcache/v3"
+	semconv "go.opentelemetry.io/collector/semconv/v1.18.0"
+)
+
+const (
+	handlerName = "aws.appsignals.UserAgentHandler"
+	// defaultTTL is how long an item in the cache will remain if it has not been re-seen.
+	defaultTTL = time.Minute
+	// cacheSize is the maximum number of unique telemetry SDK languages that can be stored before one will be evicted.
+	cacheSize = 5
+	// attrLengthLimit is the maximum length of the language and version that will be used for the user agent.
+	attrLengthLimit = 20
+
+	// TODO: Available in semconv/v1.21.0+. Replace after collector dependency is v0.91.0+.
+	attributeTelemetryDistroVersion = "telemetry.distro.version"
+)
+
+type UserAgent struct {
+	mu          sync.RWMutex
+	prebuiltStr string
+	cache       *ttlcache.Cache[string, string]
+}
+
+func NewUserAgent() *UserAgent {
+	return newUserAgent(defaultTTL)
+}
+
+func newUserAgent(ttl time.Duration) *UserAgent {
+	ua := &UserAgent{
+		cache: ttlcache.New[string, string](
+			ttlcache.WithTTL[string, string](ttl),
+			ttlcache.WithCapacity[string, string](cacheSize),
+		),
+	}
+	ua.cache.OnEviction(func(context.Context, ttlcache.EvictionReason, *ttlcache.Item[string, string]) {
+		ua.build()
+	})
+	go ua.cache.Start()
+	return ua
+}
+
+// Handler creates a named handler with the UserAgent's handle function.
+func (ua *UserAgent) Handler() request.NamedHandler {
+	return request.NamedHandler{
+		Name: handlerName,
+		Fn:   ua.handle,
+	}
+}
+
+// handle adds the pre-built user agent string to the user agent header.
+func (ua *UserAgent) handle(r *request.Request) {
+	ua.mu.RLock()
+	defer ua.mu.RUnlock()
+	request.AddToUserAgent(r, ua.prebuiltStr)
+}
+
+// Process takes the telemetry SDK language and version and adds them to the cache. If it already exists in the
```

Comment:
It might be helpful to make it clear that the key is the SDK language and the value is the version. (I don't have a lot of context with this exporter, so if it's already clear others feel free to ignore.)

Reply:
This isn't clear to me, either.  Or, rather, why it does what it does isn't clear.  Isn't this going to lead to thrashing in the cache if two producers are using the same language and different versions?  Why is the user agent of a collector exporter including details of the telemetry it is processing in the first place rather than information about itself?  The user agent shouldn't change over the process lifetime so a cache seems both unwarranted and covering up a smell.